shims/common/src/main/scala/org/apache/gluten/config/ConfigEntry.scala

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.config
+
+import org.apache.gluten.config.BackendType.BackendType
+
+trait ConfigEntry[T] {
+  def key: String
+  def doc: String
+  def version: String
+  def backend: BackendType.BackendType
+  def isPublic: Boolean
+  def alternatives: List[String]
+  def valueConverter: String => T
+  def stringConverter: T => String
+  def readFrom(conf: ConfigProvider): T
+  def defaultValue: Option[T]
+  def defaultValueString: String
+
+  final protected def readString(provider: ConfigProvider): Option[String] = {
+    alternatives.foldLeft(provider.get(key))((res, nextKey) => res.orElse(provider.get(nextKey)))
+  }
+
+  override def toString: String = {
+    s"ConfigEntry(key=$key, defaultValue=$defaultValueString, doc=$doc, " +
+      s"public=$isPublic, version=$version)"
+  }
+
+  ConfigEntry.registerEntry(this)
+}
+
+private[gluten] class OptionalConfigEntry[T](
+    _key: String,
+    _doc: String,
+    _version: String,
+    _backend: BackendType,
+    _isPublic: Boolean,
+    _alternatives: List[String],
+    _valueConverter: String => T,
+    _stringConverter: T => String)
+  extends ConfigEntry[Option[T]] {
+  override def key: String = _key
+
+  override def doc: String = _doc
+
+  override def version: String = _version
+
+  override def backend: BackendType = _backend
```

Comment:
It seems that this API is not yet called. This is designed for future use, right?

Reply:
yes.